Book overhaul - review of complete walk-through: chapters 1-6

ajmacdonald3 opened the pull request:
I've gone through chapters 1 through 6. For the most part, my edits were minor wording changes and polishing the formatting. I also updated all mapping to use sf and rnaturalearth, except I left one ggmap example at the end of chapter 6 in case folks want a colour basemap. The biggest changes were in chapter 5, where a lot of plots were not rendering properly due to not converting timestamps to POSIXct. I updated the code to convert all the timestamps to datetimes from numeric timestamps for plots and for the tabular summaries so they were more interpretable.

steffilazerte commented on vignettes/articles/05-data-cleaning.Rmd:
```diff
@@ -210,31 +209,29 @@ If you want to summarize at a finer/coarser scale, you can also create other gro
+            ts = mean(ts)) %>% 
   arrange(motusTagID, ts) %>%
-  collect() %>%
-  mutate(ts = plyr::round_any(ts, 3600),
-         ts = as_datetime(ts, tz = "UTC", origin = "1970-01-01"))
```
I'm a bit confused by you ran into errors with the timestamp. Any errors should have prevented the article from compiling, but originally it did. I think all the time stamp issues you ran into were because these two lines were omitted early on.

ajmacdonald3 replied:
Sorry, I should have given a more detailed explanation. This didn't cause an error, but summarized the data differently than the old R Book resulting in slightly different plots. If I'm interpreting it correctly, the text describes a method to summarize by `runID`, then offers alternatives such as a rounded timestamp variable (`ts.h = plyr::round_any(ts, 3600)`). In the old R Book, it doesn't actually implement this timestamp rounded to hour, it just calculates a `mean(ts)` for each `runID`, but confusingly calls that `ts.h`. I modified it to summarize by `runID`, but just keeping `mean(ts)` as `ts`, not `ts.h`. So I do think we shouldn't include `ts = plyr::round_any(ts, 3600)` because it alters the plot (see tag 16036), but I think you're right that some of the later timestamp conversions could be avoided if we kept the last line (`ts = as_datetime(ts, tz = "UTC", origin = "1970-01-01")`). So I think this is the code chunk we want:

```
df.summary <- tbl.alltags.sub %>%
  filter(tagProjID == 176, # keep only tags registered to the sample project
         !is.na(recvDeployLat) | !(recvDeployLat == 0)) %>% # drops data without lon/lat
  group_by(motusTagID, runID, recvDeployName, ambigID, 
           tagDepLon, tagDepLat, recvDeployLat, recvDeployLon) %>%
  #summarizing by runID to get max run length and mean time stamp:
  summarize(max.runLen = max(runLen), 
            ts = mean(ts)) %>% 
  arrange(motusTagID, ts) %>%
  collect() %>% 
  mutate(ts = as_datetime(ts, tz = "UTC", origin = "1970-01-01"))
```

steffilazerte commented on vignettes/articles/06-exploring-data.Rmd:
```diff
@@ -101,18 +103,21 @@ sql.motus %>%
-         recvLat, recvLon) %>% 
+         recvDeployLat, recvDeployLon) %>% 
+  collect() %>% 
+  mutate(ts = as_datetime(ts)) %>% 
```
This isn't necessary because it's already timestamp format from above

ajmacdonald3 replied:
Yes, you're right! I got a little overzealous in my timestamp conversion 😅

steffilazerte reviewed (commented):
Hi @ajmacdonald3,
Yikes, sorry for the problems in some of this, I definitely wasn't paying attention in places (05 for one!). Thank you so much for going through it so thoroughly, it makes me feel so much better :grin: And I **love** that you converted everything to sf, it's much nicer.

A couple of things: 

**Changes**  

I pushed a couple of changes that either reverted one of your changes or added content that I thought should be added. Could you check and see if you agree? Hopefully the commit messages should be clear.

**Timestamps** (the bane of my existance!)

I think the idea of timestamps could use some discussion. In general, if an article compiles (and you can see it online), it doesn't have any errors in it. So if you're running into actual errors, then there must be something that's changed, or something different on your system. I think that omitting a couple of lines at the beginning of 05-data-cleaning.R created the problem with timestamps that trickled through to the rest of that file (see my comment, where I think that change was made). In 06-exploring-data.R  the timestamp is set early on, so doesn't actually need to be set with `as_datetime()` in every step. 
  
I was a bit confused as to why you set the as_datetime() everywhere, as my preference is usually to set things once at the beginning, but I can appreciate the confusion with having `ts` sometimes be numeric and sometime date/time, which could argue for using the `as_datetime()` all the time, just in case!

So, I've thinking about this confusion and also thinking about some issues with the `getGPS()` function (which brought up the idea that sometimes it can be helpful to have the original, numeric `ts`, column). This brought me to wonder if I should go through the articles and use a column `time` when converting `ts` with `as_datetime()` and leave `ts` as numeric. Then, readers should always expect the column `time` to reflect a different kind of data than `ts`. We could then set `time = as_datetime(ts)` early on, and then only use `time = as_datetime(ts)` as required when collecting from the data base. Any plot would use `time` rather than `ts` which should tell the readers what kind of data is required. What do you think?

Also, this might be a good topic for another article on dates and times, timestamps, and databases, to clarify some of these issues for users. What do you think?

Please feel free to disagree strongly! You have the expertise with users that I lack, so I really appreciate your point of view in this!

ajmacdonald3 commented on vignettes/articles/05-data-cleaning.Rmd:
```diff
@@ -210,31 +209,29 @@ If you want to summarize at a finer/coarser scale, you can also create other gro
+ggplot(data = filter(df.summary,
                      motusTagID %in% c(16011, 16035, 16036, 16037, 16038, 16039)), 
-       aes(x = ts, y = recvDeployLat)) +
+       aes(x = as_datetime(ts), y = recvDeployLat)) +
```
We shouldn't need `as_datetime` here anymore if we reinstate in the summarizing code above. Same with any other plots relying on `df.summary` or `df.summary.sub`, I think.

ajmacdonald3 commented on vignettes/articles/05-data-cleaning.Rmd:
```diff
@@ -358,12 +365,15 @@ ggplot(data = df.plot.ambig,
+  count(motusTagID, tagDeployStart, tagDeployEnd, tagDepLat, tagDepLon) %>% 
+  collect() %>% 
+  mutate(tagDeployStart = as_datetime(tagDeployStart),
+         tagDeployEnd = as_datetime(tagDeployEnd))
```
I like your updates to use `tbl.alltags.sub` here instead of a dataframe like in the old book because it is faster/uses less memory. However, I do think we need to keep the timestamp conversions I added here because it's summarizing data pulled directly from a `tbl`, so `ts` is still numeric. There's no error if we don't convert, but the resulting summary isn't as easily interpreted by users if `ts` isn't converted to a datetime. Same for other cases where we're pulling summary information directly from `tbl`s.

ajmacdonald3 reviewed (commented):
Yes, you are correct to revert this! I started messing around when "good" was returning an empty dataframe trying to figure out what was going on, and evidently I forgot to change it back to "good" after I figured out what the issue was.

ajmacdonald3 commented:
Thanks for going through all my changes, @steffilazerte! I totally agree with the changes you pushed. Re. timestamps (yes, the bane of our existence!), I added some detail replying to your comment and added a couple of new comments that hopefully clarify things. Overall, I think you're right that I eliminated a ts to datetime conversion early on when creating the summarized data that resulted in me needing to add it back in multiple times later on. We can change that back so the summarized data has datetimes and it doesn't need to be converted each time we make a plot. However, when we're summarizing data from the `alltags` table, I think we do need to keep the conversion I added each time because we can't store `ts` as a datetime in a `tbl` (I think... correct me if I'm wrong).

I think that's a good idea to create a new column for `time` with it always in datetime format and keep `ts` as numeric. Then both options are available depending on a user's needs. Yes, an article on dates, times, and databases probably would be useful to folks - dates can be pesky to work with!

steffilazerte commented on vignettes/articles/06-exploring-data.Rmd:
```diff
@@ -101,18 +103,21 @@ sql.motus %>%
   summary()
 
 # same summary for the filtered sql data
 df.alltags.sub %>% 
   select(ts, motusTagID, runLen, speciesEN, tagDepLat, tagDepLon, 
-         recvLat, recvLon) %>% 
+         recvDeployLat, recvDeployLon) %>% 
```
Above (~ L58) the `df.alltags.sub` is created with the new fields `recvLat` and `recvLon` and `recvAlt`, which are `gps` values if we have them and `recvDeploy` values if we don't. I definitely find this confusing. Is it worth keeping that in, here? Or should I move the lines where `recvLat` etc. are created to the GPS article, and just use `recvDeployLat` etc. here and below?

Right now this errors, however, because `recvDeployLat` doesn't exist.

ajmacdonald3 replied:
My personal preference is to just use `recvDeploy` coordinates, as we have done throughout the complete walkthrough chapters. Adding in `gpsLat` and `gpsLon` fields in earlier seems like it could cause unnecessary confusion and would fit better in the GPS article.

steffilazerte replied:
I agree 100%, thanks!

steffilazerte commented:
Oh this is getting so good, I love it! 

I went through and replaced all `ts` with `time` (where it was converted). I also simplified one or two things along the way. This probably isn't the best place for major changes, but I really didn't want to forget them.

Simplifications:
- Pulling out the hour (with `hour()`), rather than rounding the hour - results in a `floor()` rather than `round()` but simpler, and I don't think it matters?
- `filter()` is smart enough to detect date/time in the standard format so you don't need to convert it with `ymd()` etc.

If you have the chance, see what you think of my edits (and my new comment, above). If you're happy, I'll go about merging and see how all of this works live. (The failing test on this branch isn't something to worry about :grin:)

Then, I'll go through the helper functions and make sure they use similar notation (but that'll be independent of this pull request).

ajmacdonald3 commented:
I took a look at your most recent changes and I think they look good! Just one more thing from me before I think we're good to go live. Stu and Denis asked me to rewrite the Acknowledgments and adjust the author/contributor list (Birds Canada as the author and all the individual people as contributors), so I'm just checking with them that my rewrite of the Acknowledgments is what they were looking for and then I'll update the Rmd file or pass the changes to you?

steffilazerte commented:
Yup to you, I'm hoping to merge and double check tomorrow, so if you think you'll get to it today then just make the changes to the branch here. Otherwise send them to me (or open a new pull request, which ever is easiest!) and we'll add them on. I'll make sure we get them in before we switch over :) Thanks again for your comments, I'm much happier with these sections now.

ajmacdonald3 commented:
Okay, I've added Stu's edits to the introduction article. Would you also be able to update the [citation](https://motuswts.github.io/motus/authors.html) to reflect the following (I wasn't sure how to update that part)?

### Authors and Citation
**Authors**

**Birds Canada**. Author, maintainer.
**John Brzustowski**. Contributor.
**Denis Lepage**. Contributor.
**Steffi LaZerte**. Contributor.
**Joey Bernard**. Contributor.
**Lucas Berrigan**. Contributor.
**Tara Crewe**. Contributor.
**Zoe Crysler**. Contributor.
**Jeremy Hussell**. Contributor.
**Catherine Jardine**. Contributor.
**Amie MacDonald**. Contributor.
**Stuart Mackenzie**. Contributor.
**Paul Morrill**. Contributor.
**Josh Sayers**. Contributor.
**Philip Taylor**. Contributor.

### Citation
Birds Canada (2022). motus: Fetch and use data from the Motus Wildlife Tracking System. https://github.com/motusWTS/motus, https://motusWTS.github.io/motus, https://motus.org/.
```
@Manual{,
  title = {motus: Fetch and use data from the Motus Wildlife Tracking System},
  author = {Birds Canada},
  year = {2022},
  note = {https://github.com/motusWTS/motus, https://motusWTS.github.io/motus, https://motus.org},
}
```
Thanks and I think we're ready to push to live! We are going to send an email out to the motus list to let folks know about the new R Book/website - I can draft the email and then if I could run it by you that would be great since you understand the updates and switch to `pkgdown` best. Thanks!! 😄

steffilazerte commented:
Great! 

Just to confirm, the citation information is for the package (in the DESCRIPTION file), not the documentation, but since documentation as well as code is part of a complete package, this makes perfect sense.  One other thing to remember is that the maintainer of a package is who most people will go to with questions etc. So as long as the email associated with Birds Canada in this context (motus@birdscanada.org) is the right email for people to contact. That should be fine. 

I'll merge everything in tomorrow, thanks again!

ajmacdonald3 commented:
Hi @steffilazerte, here is the email I've drafted to send out to Motus collaborators. Please feel free to let me know if you think anything should be changed to more accurately describe the switch to `pkgdown` and the updates to the R Book content. Thanks so much! We can send it out after you've had a chance to push to live.

---
Dear Motus collaborators,

We have recently revised the Motus R Book and created a website for the [motus R package](https://motuswts.github.io/motus/index.html). The chapters formerly found in the Motus R Book are now listed as [Articles](https://motuswts.github.io/motus/articles/index.html) on the website, and we have updated out-of-date information and added some new material. The website is automatically compiled from the R package and therefore includes function references and updates automatically when the R package is modified. The original Motus R Book will remain available as a [PDF](https://motus.org/motusRBook/archives/MotusRBook2018-01.pdf).

We hope you find this updated material helpful. We welcome contributions and suggestions from the Motus community to both the articles and the motus R package. If you would like to make a contribution or suggestion, please submit a pull request or issue on [GitHub](https://github.com/MotusWTS/motus) or email us at [motus@birdscanada.org](mailto:motus@birdscanada.org). As always, please don’t hesitate to contact us if you have any questions.

Sincerely,

The Motus Team

steffilazerte commented:
Looks good! Only suggestion I would have is that you list "open an issue", before submitting a pull request, as ideally for anything bigger than a typo fix, there would be an issue first. Also consider adding a link to "issue" to https://github.com/motuswts/motus/issues, although hmm, I see we don't have issues active for this project. We might want to change that. Thoughts? (unfortunately today got away from me so I won't be putting this together until Monday, sorry about that!)

ajmacdonald3 commented:
Great, I will adjust to suggest issue before pull request. Yes, I noticed we didn't have issues active for the motus repository. I think we should activate them and I can adjust the settings if you agree. No worries, I was mostly caught up working on other things too and Monday is probably a better time to send out an email than Friday afternoon anyway :)
